File: connectors/connector-basecamp/src/main/java/com/taskadapter/connector/basecamp/BasecampFactory.java
package com.taskadapter.connector.basecamp;

import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.taskadapter.connector.basecamp.transport.BaseCommunicator;
import com.taskadapter.connector.basecamp.transport.ObjectAPIFactory;
import com.taskadapter.connector.common.ConfigUtils;
import com.taskadapter.connector.definition.Descriptor;
import com.taskadapter.connector.definition.PluginFactory;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.model.Field;

import java.util.List;

public class BasecampFactory implements PluginFactory<BasecampConfig, WebConnectorSetup> {

    private static final Descriptor DESCRIPTOR = new Descriptor(BasecampConnector.ID, "Basecamp 2");

    private final ObjectAPIFactory factory = new ObjectAPIFactory(
            new BaseCommunicator());

    @Override
    public List<Field<?>> getAllFields() {
        return BasecampField.fields;
    }

    @Override
    public List<Field<?>> getDefaultFieldsForNewConfig() {
        return BasecampField.fields;
    }

    @Override
    public BasecampConnector createConnector(BasecampConfig config, WebConnectorSetup setup) {
        return new BasecampConnector(config, setup, factory);
    }

    @Override
    public Descriptor getDescriptor() {
        return DESCRIPTOR;
    }

    @Override
    public JsonElement writeConfig(BasecampConfig config) {
        return ConfigUtils.createDefaultGson().toJsonTree(config);
    }

    @Override
    public BasecampConfig readConfig(JsonElement config)
            throws JsonParseException {
        return ConfigUtils.createDefaultGson().fromJson(config, BasecampConfig.class);
    }

    @Override
    public BasecampConfig createDefaultConfig() {
        return new BasecampConfig();
    }
}

File: connectors/connector-basecamp/src/main/scala/com/taskadapter/connector/basecamp/BasecampFactory.scala
package com.taskadapter.connector.basecamp

import com.google.gson.{JsonElement, JsonParseException}
import com.taskadapter.connector.basecamp.transport.{BaseCommunicator, ObjectAPIFactory}
import com.taskadapter.connector.common.ConfigUtils
import com.taskadapter.connector.definition.{Descriptor, PluginFactory, WebConnectorSetup}
import com.taskadapter.model.Field

class BasecampFactory extends PluginFactory[BasecampConfig, WebConnectorSetup] {
  val DESCRIPTOR = Descriptor(BasecampConnector.ID, "Basecamp 2")
  final private val factory = new ObjectAPIFactory(new BaseCommunicator)

  override def createConnector(config: BasecampConfig, setup: WebConnectorSetup) = new BasecampConnector(config, setup, factory)

  override def getSuggestedCombinations: Seq[Field[_]] = BasecampField.fields

  override def getDescriptor = DESCRIPTOR

  override def writeConfig(config: BasecampConfig): JsonElement = ConfigUtils.createDefaultGson.toJsonTree(config)

  @throws[JsonParseException]
  override def readConfig(config: JsonElement): BasecampConfig =
    ConfigUtils.createDefaultGson.fromJson(config, classOf[BasecampConfig])

  override def createDefaultConfig = new BasecampConfig
}